### PR TITLE
Grad handling

### LINKED ISSUE #602: Enable emulator predictions with grad
Requirement for #521 

Currently emulators return predictions without gradient information. Often the user just wants the prediction so this felt appropriate, but for some downstream tasks (e.g., MCMC) that prediction feeds into a computation that needs to be differentiated. 

A solution would be to return predictions with gradient information and any downstream code handle it. Another option would be to include a `with_grad` option in `predict()` to enable both. Is there another option? Which is better?

### PR DESCRIPTION
**Pull request by radka-j**
Closes #602

TODO:
- [ ] transformed emulator
- [ ] ensembles

### CODE REVIEW

File: autoemulate/experimental/emulators/transformed/base.py [hunk 1]
```diff
@@ -367,14 +364,24 @@ def _fit(self, x: TensorLike, y: TensorLike):
...
+                return self._inv_transform_y_tensor(y_t_pred)
+        else:
+            with torch.no_grad():
+                # Transform and invert transform for prediction in original data space
+                x_t = self._transform_x(x)
+                y_t_pred = self.model.predict(x_t, with_grad)
+
+                # If TensorLike, transform tensor back to original space
+                if isinstance(y_t_pred, TensorLike):
+                    return self._inv_transform_y_tensor(y_t_pred)
```
Comment (sgreenbury) [inline autoemulate/experimental/emulators/transformed/base.py:384]:
```suggestion
        with torch.set_grad_enabled(with_grad):
            # Transform and invert transform for prediction in original data space
            x_t = self._transform_x(x)
            y_t_pred = self.model.predict(x_t, with_grad)

            # If TensorLike, transform tensor back to original space
            if isinstance(y_t_pred, TensorLike):
                return self._inv_transform_y_tensor(y_t_pred)
```

File: autoemulate/experimental/emulators/base.py [hunk 2]
```diff
@@ -51,16 +48,16 @@ def model_name(cls) -> str:
         return cls.__name__
 
     @abstractmethod
-    def _predict(self, x: TensorLike) -> OutputLike:
+    def _predict(self, x: TensorLike, with_grad: bool = False) -> OutputLike:
```
Comment (sgreenbury) [inline autoemulate/experimental/emulators/base.py:51]: Given there is the option to set_grad_enabled, perhaps we do not need to add the the `_predict()` API? This might be helpful so that subclassing would not need to consider this by default.

**Review by sgreenbury — COMMENTED** [review]
I think this is looking great, thanks @radka-j!

My main thought is whether we might be able to refactor to only pass `with_grad` in the `predict()` public API and wrap the call within the default implementation with a context manager, what do you think? E.g.:
```python
def predict(self, x, with_grad):
    # ...
    # We could possibly add a new method to check if instance supports gradients?
    if not self.supports_grad():
        raise ...
    
    with torch.set_grad_enabled(with_grad):
        return self._predict(x)
```

**Comment by radka-j** [comment]
@sgreenbury I updated the code to use `with torch.set_grad_enabled(with_grad):`, it looks much better now, thanks!! I had to use it inside the `_predict()` method rather than at the top level `predict()` because of the ensemble and transformed emulators.

**Review by sgreenbury — APPROVED** [review]

File: autoemulate/experimental/emulators/transformed/base.py [hunk 3]
```diff
@@ -137,6 +133,11 @@ def __init__(  # noqa: PLR0913
         self.n_samples = n_samples
         self.full_covariance = full_covariance and y.shape[1] <= max_targets
         TorchDeviceMixin.__init__(self, device=device)
+        self.supports_grad = (
+            self.model.supports_grad
+            and all(t.bijective for t in self.x_transforms)
+            and all(t.bijective for t in self.y_transforms)
```
[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/emulators/transformed/base.py:139]: Will do
